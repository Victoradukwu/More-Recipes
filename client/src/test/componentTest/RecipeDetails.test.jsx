import React from 'react';
import { shallow, mount } from 'enzyme';
import {
  RecipeDetailsPage,
  mapStateToProps,
  mapDispatchToProps
} from '../../components/recipes/recipeDetails/RecipeDetailsPage';

const props = {
  match: {
    params: { id: 1 }
  },
  getSingleRecipe: jest.fn(),
  history: {
    push: jest.fn()
  },
  upvoteRecipe: jest.fn(),
  downvoteRecipe: jest.fn(),
  reviewRecipe: jest.fn(),
  favoriteRecipe: jest.fn(),
  commentChange: jest.fn()
};

describe('RecipeDetailsPage', () => {
  it('matches snapshot', () => {
    const wrapper = shallow(<RecipeDetailsPage {...props} />);
    expect(wrapper).toMatchSnapshot();
  });

  describe('instance and lifecycle methods', () => {
    describe('handeleUpvote method', () => {
      it('calls the upvoteRecipe prop function', () => {
        const wrapper = shallow(<RecipeDetailsPage {...props} />);
        wrapper.instance().handleUpvote();
        expect(props.upvoteRecipe).toHaveBeenCalled();
      });
    });

    describe('handeleDownvote method', () => {
      it('calls the downvoteRecipe prop function', () => {
        const wrapper = shallow(<RecipeDetailsPage {...props} />);
        wrapper.instance().handleDownvote();
        expect(props.downvoteRecipe).toHaveBeenCalled();
      });
    });

    describe('handeleFavorite method', () => {
      it('calls the favoriteRecipe prop function', () => {
        const wrapper = shallow(<RecipeDetailsPage {...props} />);
        wrapper.instance().handleFavorite();
        expect(props.favoriteRecipe).toHaveBeenCalled();
      });
    });

    it('handle comment change', () => {
      const value = 'Awesome';
      const name = 'comment';
      const wrapper = mount(<RecipeDetailsPage {...props} />);
      wrapper.find('textarea').simulate('change', {
        target: { name, value },
        preventDefault: () => {}
      });
      expect(wrapper.state('comment')).toBe(value);
    });


    describe('handleReview method', () => {
      it('calls reviewRecipe props function', () => {
        const eventMock = {
          preventDefault: jest.fn()
        };
        const wrapper = shallow(<RecipeDetailsPage {...props} />);
        wrapper.instance().handleReview(eventMock);
        expect(props.reviewRecipe).toHaveBeenCalled();
      });

      it('calls setState with the comment from the component state', () => {
        const eventMock = {
          preventDefault: jest.fn()
        };
        const wrapper = shallow(<RecipeDetailsPage {...props} />);
        const setState = jest.spyOn(wrapper.instance(), 'setState');
        wrapper.instance().handleReview(eventMock);
        expect(setState).toHaveBeenCalledWith({ comment: '' });
      });
    });
  });

  describe('conatiner functions', () => {
    test('mapStateToProps', () => {
      expect(mapStateToProps({ user: {} })).toHaveProperty('recipe');
      expect(mapStateToProps({ user: {} })).toBeInstanceOf(Object);
    });

    it('mapDispatchToProps', () => {
      const dispatch = jest.fn();
      expect(mapDispatchToProps(dispatch)).toHaveProperty('upvoteRecipe');
      expect(mapDispatchToProps(dispatch)).toHaveProperty('downvoteRecipe');
      expect(mapDispatchToProps(dispatch)).toHaveProperty('favoriteRecipe');
      expect(mapDispatchToProps(dispatch)).toHaveProperty('reviewRecipe');
      expect(mapDispatchToProps(dispatch)).toHaveProperty('getSingleRecipe');
      expect(mapDispatchToProps(dispatch)).toBeInstanceOf(Object);

      const { upvoteRecipe } = mapDispatchToProps(dispatch);
      upvoteRecipe();
      expect(dispatch).toHaveBeenCalled();
    });
  });
});
