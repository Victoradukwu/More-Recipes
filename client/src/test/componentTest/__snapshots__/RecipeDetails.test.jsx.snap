// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RecipeDetailsPage Snapshot matches snapshot 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RecipeDetailsPage
    commentChange={[MockFunction]}
    downvoteRecipe={[MockFunction]}
    favoriteRecipe={[MockFunction]}
    getSingleRecipe={
      [MockFunction] {
        "calls": Array [
          Array [
            1,
          ],
        ],
      }
    }
    history={
      Object {
        "push": [MockFunction] {
          "calls": Array [
            Array [
              "/signin",
            ],
          ],
        },
      }
    }
    loggedInUser={1}
    match={
      Object {
        "params": Object {
          "id": 1,
        },
      }
    }
    recipe={Object {}}
    reviewRecipe={[MockFunction]}
    upvoteRecipe={[MockFunction]}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <br />,
        <br />,
        <div
          className="row"
        >
          <div
            className="col-sm-2"
          />
          <div
            className="col-sm-8"
          >
            <Ratings
              className="btn btn-lg fa fa-lg fa-heart"
              downvoteRecipe={[Function]}
              favoriteRecipe={[Function]}
              recipe={Object {}}
              upvoteRecipe={[Function]}
            />
            <RecipeInfo
              recipe={Object {}}
            />
            <ReviewForm
              addReview={[Function]}
              comment=""
              handleCommentChange={[Function]}
              recipeId={1}
            />
          </div>
          <div
            className="col-sm-2"
          />
        </div>,
        <br />,
        <br />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "br",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "br",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <div
              className="col-sm-2"
            />,
            <div
              className="col-sm-8"
            >
              <Ratings
                className="btn btn-lg fa fa-lg fa-heart"
                downvoteRecipe={[Function]}
                favoriteRecipe={[Function]}
                recipe={Object {}}
                upvoteRecipe={[Function]}
              />
              <RecipeInfo
                recipe={Object {}}
              />
              <ReviewForm
                addReview={[Function]}
                comment=""
                handleCommentChange={[Function]}
                recipeId={1}
              />
            </div>,
            <div
              className="col-sm-2"
            />,
          ],
          "className": "row",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "col-sm-2",
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": Array [
                <Ratings
                  className="btn btn-lg fa fa-lg fa-heart"
                  downvoteRecipe={[Function]}
                  favoriteRecipe={[Function]}
                  recipe={Object {}}
                  upvoteRecipe={[Function]}
                />,
                <RecipeInfo
                  recipe={Object {}}
                />,
                <ReviewForm
                  addReview={[Function]}
                  comment=""
                  handleCommentChange={[Function]}
                  recipeId={1}
                />,
                undefined,
              ],
              "className": "col-sm-8",
            },
            "ref": null,
            "rendered": Array [
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "className": "btn btn-lg fa fa-lg fa-heart",
                  "downvoteRecipe": [Function],
                  "favoriteRecipe": [Function],
                  "recipe": Object {},
                  "upvoteRecipe": [Function],
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "recipe": Object {},
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              Object {
                "instance": null,
                "key": undefined,
                "nodeType": "function",
                "props": Object {
                  "addReview": [Function],
                  "comment": "",
                  "handleCommentChange": [Function],
                  "recipeId": 1,
                },
                "ref": null,
                "rendered": null,
                "type": [Function],
              },
              undefined,
            ],
            "type": "div",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "col-sm-2",
            },
            "ref": null,
            "rendered": null,
            "type": "div",
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "br",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {},
        "ref": null,
        "rendered": null,
        "type": "br",
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <br />,
          <br />,
          <div
            className="row"
          >
            <div
              className="col-sm-2"
            />
            <div
              className="col-sm-8"
            >
              <Ratings
                className="btn btn-lg fa fa-lg fa-heart"
                downvoteRecipe={[Function]}
                favoriteRecipe={[Function]}
                recipe={Object {}}
                upvoteRecipe={[Function]}
              />
              <RecipeInfo
                recipe={Object {}}
              />
              <ReviewForm
                addReview={[Function]}
                comment=""
                handleCommentChange={[Function]}
                recipeId={1}
              />
            </div>
            <div
              className="col-sm-2"
            />
          </div>,
          <br />,
          <br />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <div
                className="col-sm-2"
              />,
              <div
                className="col-sm-8"
              >
                <Ratings
                  className="btn btn-lg fa fa-lg fa-heart"
                  downvoteRecipe={[Function]}
                  favoriteRecipe={[Function]}
                  recipe={Object {}}
                  upvoteRecipe={[Function]}
                />
                <RecipeInfo
                  recipe={Object {}}
                />
                <ReviewForm
                  addReview={[Function]}
                  comment=""
                  handleCommentChange={[Function]}
                  recipeId={1}
                />
              </div>,
              <div
                className="col-sm-2"
              />,
            ],
            "className": "row",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "col-sm-2",
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": Array [
                  <Ratings
                    className="btn btn-lg fa fa-lg fa-heart"
                    downvoteRecipe={[Function]}
                    favoriteRecipe={[Function]}
                    recipe={Object {}}
                    upvoteRecipe={[Function]}
                  />,
                  <RecipeInfo
                    recipe={Object {}}
                  />,
                  <ReviewForm
                    addReview={[Function]}
                    comment=""
                    handleCommentChange={[Function]}
                    recipeId={1}
                  />,
                  undefined,
                ],
                "className": "col-sm-8",
              },
              "ref": null,
              "rendered": Array [
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "className": "btn btn-lg fa fa-lg fa-heart",
                    "downvoteRecipe": [Function],
                    "favoriteRecipe": [Function],
                    "recipe": Object {},
                    "upvoteRecipe": [Function],
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "recipe": Object {},
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                Object {
                  "instance": null,
                  "key": undefined,
                  "nodeType": "function",
                  "props": Object {
                    "addReview": [Function],
                    "comment": "",
                    "handleCommentChange": [Function],
                    "recipeId": 1,
                  },
                  "ref": null,
                  "rendered": null,
                  "type": [Function],
                },
                undefined,
              ],
              "type": "div",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "col-sm-2",
              },
              "ref": null,
              "rendered": null,
              "type": "div",
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {},
          "ref": null,
          "rendered": null,
          "type": "br",
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
